cpp/src/arrow/util/async_generator.h

```diff
@@ -1120,43 +1272,44 @@ class MergedGenerator {
 
   struct OuterCallback {
     void operator()(const Result<AsyncGenerator<T>>& maybe_next) {
-      bool should_purge = false;
       bool should_continue = false;
+      bool should_mark_gen_complete = false;
```

Comment:
this is actually something like `all_requests_fulfilled`?

Reply:
The other variables are named based on "action to take" and not "state that we were in".  The logic is roughly...

```
bool should_do_x = false;
bool should_do_y = false;
with lock:
  alter_state()
  should_do_x = in_state_a
  should_do_y = in_state_b
if should_do_x:
  ...
if should_do_y:
  ...
```

So I think `should_mark_gen_complete` is consistent.